### PR TITLE
Lower derived type.

### PR DESCRIPTION
**Pull request by schweitzpgi**
Simple preliminary type lowering.

### CODE REVIEW

File: flang/lib/Lower/ConvertType.cpp [hunk 1]
```diff
@@ -221,12 +223,22 @@ struct TypeBuilder {
...
+        for (auto iter = components.begin(), iend = components.end();
+             iter != iend; ++iter) {
+          auto &field = *iter;
+          cs.emplace_back(field.name().ToString(), genSymbolType(field));
+        }
+        for (auto iter : tySpec->parameters())
+          if (iter.second.isLen()) {
+            ps.emplace_back(
+                iter.first.ToString(),
+                /* FIXME - is the type available? Fake with index type. */
```
Comment (jeanPerier) [inline flang/lib/Lower/ConvertType.cpp:236]: `DerivedTypeSpec` represents an instance of of a derived type. It does not store the types of the type parameters, but their values (assumed, deferred, explicit expr). The type information is stored in the Symbol that represents the type definition. You can get the type definition from a `DerivedTypeSpec` with the `typeSymbol()` member function.
All that to say, I think you can use something like:
```
for (const auto& param: Fortran::semantics::OrderParameterDeclarations(tySpec->typeSymbol())
   if (param.get<TypeParamDetails>().attr() == Fortran::common::TypeParamAttr::Len))
       ps.emplace_back(param.name.ToString(), genSymbolType(param));
      
```

See `OrderParameterDeclarations` declaration [here ](https://github.com/llvm/llvm-project/blob/ebe74d9592d097501f376c2086e58d35aa318896/flang/include/flang/Semantics/tools.h#L207).

File: flang/lib/Lower/ConvertType.cpp [hunk 2]
```diff
@@ -221,12 +223,22 @@ struct TypeBuilder {
         std::vector<std::pair<std::string, mlir::Type>> cs;
         auto &symbol = tySpec->typeSymbol();
         auto rec = fir::RecordType::get(context, toStringRef(symbol.name()));
```
Comment (jeanPerier) [inline flang/lib/Lower/ConvertType.cpp:225]: Question here: should derived type name be mangled, especially to include the kind parameters values ? Not asking to do this know, I am just wondering about the TODOs remaining in this code.

File: flang/lib/Lower/ConvertExpr.cpp [hunk 3]
```diff
@@ -957,20 +957,25 @@ class ScalarExprLowering {
...
+      auto recTy = ty.cast<fir::RecordType>();
       sym = &field->GetLastSymbol();
-      auto name = sym->name().ToString();
-      // FIXME: as we're walking the chain of field names, we need to update the
-      // subtype as we drill down
-      coorArgs.push_back(builder.create<fir::FieldIndexOp>(getLoc(), name, ty));
+      auto name = toStringRef(sym->name());
+      auto fldTy = recTy.getType(name);
+      coorArgs.push_back(builder.create<fir::FieldIndexOp>(
+          loc, fldTy, name, mlir::TypeAttr::get(recTy),
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertExpr.cpp:971]: Thanks. Yes, that's a bug.

File: flang/lib/Lower/ConvertExpr.cpp [hunk 4]
```diff
@@ -957,20 +957,25 @@ class ScalarExprLowering {
...
+          /*lenparams=*/mlir::ValueRange{}));
+      ty = fldTy;
     }
     assert(sym && "no component(s)?");
     ty = builder.getRefType(ty);
-    return builder.create<fir::CoordinateOp>(getLoc(), ty, obj, coorArgs);
+    return builder.create<fir::CoordinateOp>(loc, ty, fir::getBase(obj),
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertExpr.cpp:977]: Yes. I was intentionally ignoring that for the moment.

File: flang/lib/Lower/ConvertType.cpp (hunk 2 above)
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertType.cpp:225]: They need to be. The internal name mangling scheme has a solution for that.

File: flang/lib/Lower/ConvertType.cpp [hunk 5]
```diff
@@ -221,12 +223,22 @@ struct TypeBuilder {
...
-        // TODO: use Fortran::semantics::ComponentIterator to go through
-        // components. or use similar mechanism. We probably want to go through
-        // the Ordered components.
-        TODO("lower derived type to fir types");
+        Fortran::semantics::OrderedComponentIterator components(*tySpec);
+        for (auto iter = components.begin(), iend = components.end();
+             iter != iend; ++iter) {
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertType.cpp:228]: Sounds good. I'll tweak this.

File: flang/lib/Lower/ConvertType.cpp [hunk 6]
```diff
@@ -221,12 +223,22 @@ struct TypeBuilder {
...
+          auto &field = *iter;
+          cs.emplace_back(field.name().ToString(), genSymbolType(field));
```
[the remark this answers is not in the document]
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertType.cpp:230]: I'll note that here in a comment.

File: flang/lib/Lower/ConvertType.cpp (hunk 1 above)
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertType.cpp:236]: Thanks!

File: flang/lib/Lower/ConvertType.cpp (hunk 2 above)
Comment (schweitzpgi) [reply flang/lib/Lower/ConvertType.cpp:227]: I added a comment here.

**Review by jeanPerier — APPROVED** [review]

**Pull request merged**